### PR TITLE
feat(core): add support for managing failed records in rsources.JobService

### PR DESCRIPTION
**Pull request by Sidddddarth**
# Description

rsources.JobService now provides support for adding/deleting and getting failed records.

## Notion Ticket

[Failed records support](https://www.notion.so/rudderstacks/Failed-records-support-in-rsources-JobService-6a44ebf12aba4fd7b5b9099d334fd28a)

## Security

- [x] The code changed/added as part of this pull request won't create any security issues with how the software is being used.

### CODE REVIEW

File: services/rsources/rsources.go [hunk 1]
```diff
@@ -101,7 +111,7 @@ type JobService interface {
 	AddFailedRecords(ctx context.Context, tx *sql.Tx, jobRunId string, key JobTargetKey, records []json.RawMessage) error
 
 	// TODO: future extension
```
Comment (atzoum) [inline services/rsources/rsources.go:113]: need to adapt the doc here as well

File: services/rsources/handler.go [hunk 2]
```diff
@@ -106,17 +106,91 @@ func (*sourcesHandler) IncrementStats(ctx context.Context, tx *sql.Tx, jobRunId
 	return err
 }
 
-func (*sourcesHandler) AddFailedRecords(_ context.Context, _ *sql.Tx, _ string, _ JobTargetKey, _ []json.RawMessage) error {
+func (*sourcesHandler) AddFailedRecords(ctx context.Context, tx *sql.Tx, jobRunId string, key JobTargetKey, records []json.RawMessage) error {
+	sqlStatement := `insert into "failed_keys" (
```
Comment (atzoum) [inline services/rsources/handler.go:110]: using a prepared statement here could be beneficial as well

File: services/rsources/handler.go [hunk 3]
```diff
@@ -106,17 +106,91 @@ func (*sourcesHandler) IncrementStats(ctx context.Context, tx *sql.Tx, jobRunId
...
+	}
+	if err := rows.Err(); err != nil {
+		return FailedRecords{}, err
+	}
+
+	return failedRecords, nil
 }
 
 func (sh *sourcesHandler) Delete(ctx context.Context, jobRunId string) error {
 	sqlStatement := `delete from "rsources_stats" where job_run_id = $1`
```
Comment (atzoum) [inline services/rsources/handler.go:185]: should perform both statements as an atomic action in the same transaction

File: services/rsources/handler.go [hunk 4]
```diff
@@ -106,17 +106,91 @@ func (*sourcesHandler) IncrementStats(ctx context.Context, tx *sql.Tx, jobRunId
...
 	_, err := sh.localDB.ExecContext(ctx, sqlStatement, jobRunId)
+	if err != nil {
+		return err
+	}
+
+	sqlStatement = `delete from "failed_keys" where job_run_id = $1`
+	_, err = sh.localDB.ExecContext(ctx, sqlStatement, jobRunId)
+
 	return err
 }
```
Comment (atzoum) [inline services/rsources/handler.go:196]: do we need `DropStats` method?

File: services/rsources/handler.go [hunk 5]
```diff
@@ -106,17 +106,91 @@ func (*sourcesHandler) IncrementStats(ctx context.Context, tx *sql.Tx, jobRunId
...
 
 func (sh *sourcesHandler) Delete(ctx context.Context, jobRunId string) error {
 	sqlStatement := `delete from "rsources_stats" where job_run_id = $1`
 	_, err := sh.localDB.ExecContext(ctx, sqlStatement, jobRunId)
+	if err != nil {
+		return err
+	}
+
+	sqlStatement = `delete from "failed_keys" where job_run_id = $1`
```
Comment (atzoum) [inline services/rsources/handler.go:191]: cleanup loop should be adapted as well. We don't want to keep old failed keys

File: services/rsources/handler_test.go [hunk 6]
```diff
@@ -63,6 +64,38 @@ var _ = Describe("Using sources handler", func() {
...
+				[]byte(`{"record-2": "id-2"}`),
+			})
+			jobFilters := JobFilter{
+				SourceID:  []string{"source_id"},
+				TaskRunID: []string{"task_run_id"},
+			}
+
+			failedRecords, err := sh.GetFailedRecords(context.Background(), jobRunId, jobFilters)
+			Expect(err).NotTo(HaveOccurred(), "it should be able to get failed records")
+			expcetdRecords := FailedRecords{
```
Comment (atzoum) [inline services/rsources/handler_test.go:80]: small typo
```suggestion
			expectedRecords := FailedRecords{
```

File: services/rsources/handler.go [hunk 7]
```diff
@@ -241,5 +388,12 @@ func (sh *sourcesHandler) setupLogicalReplication(ctx context.Context) error {
 			return fmt.Errorf("failed to create subscription on shared database: %w", err)
 		}
 	}
+
+	refreshSubscriptionQuery := fmt.Sprintf(`ALTER SUBSCRIPTION %s REFRESH PUBLICATION`, subscriptionName)
```
Comment (atzoum) [inline services/rsources/handler.go:392]: 👍

**Review by atzoum — APPROVED** [review]

File: services/rsources/handler.go [hunk 8]
```diff
@@ -106,24 +106,107 @@ func (*sourcesHandler) IncrementStats(ctx context.Context, tx *sql.Tx, jobRunId
...
+	filterParams := []interface{}{}
+	filters := `WHERE job_run_id = $1`
+	filterParams = append(filterParams, jobRunId)
+
+	if len(filter.TaskRunID) > 0 {
+		filters += fmt.Sprintf(` AND task_run_id  = ANY ($%d)`, len(filterParams)+1)
+		filterParams = append(filterParams, pq.Array(filter.TaskRunID))
+	}
+
+	if len(filter.SourceID) > 0 {
```
Comment (cisse21) [inline services/rsources/handler.go:148]: What is the use case where we might have to filter based on multiple sourceID's?

Comment (atzoum) [reply services/rsources/handler.go:148]: It is part of the agreed contract. I guess, sourceID might change when a job is reconfigured

File: services/rsources/handler.go [hunk 9]
```diff
@@ -106,24 +106,107 @@ func (*sourcesHandler) IncrementStats(ctx context.Context, tx *sql.Tx, jobRunId
 	return err
 }
 
-func (*sourcesHandler) AddFailedRecords(_ context.Context, _ *sql.Tx, _ string, _ JobTargetKey, _ []json.RawMessage) error {
-	return nil
+func (*sourcesHandler) AddFailedRecords(ctx context.Context, tx *sql.Tx, jobRunId string, key JobTargetKey, records []json.RawMessage) (err error) {
+	stmt, err := tx.Prepare(`insert into "rsources_failed_keys" (
```
Comment (lvrach) [inline services/rsources/handler.go:110]: I am wondering if an insert with multiple rows would be a better choice here, both in terms of performance and consistency (no partial failures).

Depending on the domain partial failures might not be appropriate

**Review by lvrach — APPROVED** [review]

File: services/rsources/handler.go (hunk 9 above)
Comment (atzoum) [reply services/rsources/handler.go:110]: `pq` doesn't appear to have support for multi-row insert, plus we will also need to cater for the limit of 1000 rows per multi-row insert...
That being said, I don't believe it is worth the trouble

**Pull request merged**